tests/neg-custom-args/isInstanceOf/enum-approx2.scala

```diff
@@ -0,0 +1,9 @@
+sealed trait Exp[T]
+case class Fun[A, B](f: Exp[A => B]) extends Exp[A => B]
+
+class Test {
+  def eval[T](e: Exp[T]) = e match {
+    case Fun(x: Fun[Int, Double]) => ???          // error
+    case Fun(x: Exp[Int => String]) => ???        // error
+  }
```

Comment:
Looking at this with @liufengyun, we realize this complication has to do with compiling a match like:

```scala
def eval[T](e: Exp[T]) = e match {
  case f @ Fun(x: Fun[Int, Double]) => ???
}
```

The correct desugaring is a bit surprising to me, because we want to ensure `f: Fun[Int, Double]`, but this is not justified where `f` is bound but only after the inner match.

```scala
def eval[T](e: Exp[T]) = e match {
  case f1: Fun[a1, b1] =>
    f1.f match {
      case f2: Fun[Int, Double] =>
        //***Here*** we learn that a1 = Int and b1 = Double
        //So only ***here*** we can bind `f` with the correct type.
        val f: Fun[Int, Double] = f1
    }
}
```

We suspect the typechecker currently refines type variables `a1` and `b1` too soon, which has been fine until now, but strictly speaking means (I think) the current typechecker output is not well-typed in a sound typesystem.